### PR TITLE
Migrate `lexical-mark` to TypeScript

### PR DESCRIPTION
**Pull request by maasencioh**
Help on #1242

### CODE REVIEW

File: packages/lexical-mark/src/MarkNode.ts [hunk 1]
```diff
@@ -168,6 +170,6 @@ export function $createMarkNode(ids: Array<string>): MarkNode {
   return new MarkNode(ids);
 }
 
-export function $isMarkNode(node: ?LexicalNode): boolean %checks {
+export function $isMarkNode(node?: LexicalNode | null): node is MarkNode {
```
Comment (thegreatercurve) [inline packages/lexical-mark/src/MarkNode.ts:173]: We want to make sure we pass something to this method: 
```suggestion
export function $isMarkNode(node: LexicalNode | null): node is MarkNode {
```

**Review by thegreatercurve — COMMENTED** [review]
LGTM but I would leave out the `shared` packages migration, at least for the moment. 

I had issues trying to get those to work interoperably between Flow and TS in #2135, hence the failing build here as well. It was easier just to create a temporary and separate `shared-ts` package which we use in the TypeScript files. 

My idea is to remove `shared` and rename `shared-ts` back to `shared` once everything has been migrated.

**Comment by thegreatercurve** [comment]
Thanks for helping to work on this @maasencioh! 🙏

**Comment by maasencioh** [comment]
> My idea is to remove shared and rename shared-ts back to shared once everything has been migrated.
Thanks I just saw your PR, so for the moment I'll revert that package

**Comment by thegreatercurve** [comment]
I think you need to update this line in the build script as well to `.ts`: https://github.com/facebook/lexical/blob/main/scripts/build.js#L541

**Comment by maasencioh** [comment]
@thegreatercurve thank you, sorry, a lot of new learning about how to do it, but after, I'm pretty sure to be able to do it with fewer fix commits 😅

**Review by thegreatercurve — APPROVED** [review]

**Comment by thegreatercurve** [comment]
> @thegreatercurve thank you, sorry, a lot of new learning about how to do it, but I'm pretty sure to be able to do it with fewer fix commits next time 😅

@maasencioh It's really no problem at all! All of the commits on a PR get squashed on merge any way, so feel free to add as many as you like.

**Pull request merged**